p4include/psa.p4

```diff
@@ -25,403 +24,518 @@ limitations under the License.
 const   PortId_t         PORT_CPU = unspecified;
+// END:Type_defns
+
 #endif
 
-#include "core.p4"
+// BEGIN:Metadata_types
+enum PacketPath_t {
+    NORMAL,     /// Packet received by ingress that is none of the cases below.
+    NORMAL_UNICAST,   /// Normal packet received by egress which is unicast
+    NORMAL_MULTICAST, /// Normal packet received by egress which is multicast
+    CLONE_I2E,  /// Packet created via a clone operation in ingress,
+                /// destined for egress
+    CLONE_E2E,  /// Packet created via a clone operation in egress,
+                /// destined for egress
+    RESUBMIT,   /// Packet arrival is the result of a resubmit operation
+    RECIRCULATE /// Packet arrival is the result of a recirculate operation
+}
 
-/* --------------------------------------------------------------------- *
-/**
- * PSA supported metadata types
- */
-struct psa_parser_input_metadata_t {
+struct psa_ingress_parser_input_metadata_t {
   PortId_t                 ingress_port;
-  InstanceType_t           instance_type;
+  PacketPath_t             packet_path;
+}
+
+struct psa_egress_parser_input_metadata_t {
+  PortId_t                 egress_port;
+  PacketPath_t             packet_path;
+}
+
+struct psa_parser_output_metadata_t {
+  ParserError_t            parser_error;
 }
 
 struct psa_ingress_input_metadata_t {
+  // All of these values are initialized by the architecture before
+  // the Ingress control block begins executing.
   PortId_t                 ingress_port;
-  InstanceType_t           instance_type;  /// Clone or Normal
-  /// set by the runtime in the parser, these are not under programmer control
-  ParserStatus_t           parser_status;
-  ParserErrorLocation_t    parser_error_location;
+  PacketPath_t             packet_path;
+  Timestamp_t              ingress_timestamp;
+  ParserError_t            parser_error;
 }
-
+// BEGIN:Metadata_ingress_output
 struct psa_ingress_output_metadata_t {
-  PortId_t                 egress_port;
+  // The comment after each field specifies its initial value when the
+  // Ingress control block begins executing.
+  ClassOfService_t         class_of_service; // 0
+  bool                     clone;            // false
+  PortId_t                 clone_port;       // undefined
+  ClassOfService_t         clone_class_of_service; // 0
+  bool                     drop;             // true
+  bool                     resubmit;         // false
+  MulticastGroup_t         multicast_group;  // 0
+  PortId_t                 egress_port;      // undefined
+  bool                     truncate;         // false
+  PacketLength_t           truncate_payload_bytes;  // undefined
 }
-
+// END:Metadata_ingress_output
 struct psa_egress_input_metadata_t {
+  ClassOfService_t         class_of_service;
   PortId_t                 egress_port;
-  InstanceType_t           instance_type;  /// Clone or Normal
+  PacketPath_t             packet_path;
```

Comment:
If by "PRE is not defined so far" you mean "not so far yet in this file", then agreed, but even once it is defined in this file, these metadata fields aren't mentioned there.  The metadata fields are only defined by the architecture to affect the PRE's behavior, but not via PRE method calls, since it has none that can be called from the data plane (as of now, the PRE and BQE externs only exist in the psa.p4 file to give an instance name for the control plane API to use when making calls to modify their configurations).

By a link to the PSA spec, do you mean "See Section X.Y of the PSA specification", or you mean a URL?

Reply:
You could at least say "packet replication engine" someplace.

Reply:
Oh, you meant "PRE" on the following line after your comment.  Changing that to PacketReplicationEngine, as it was the only occurrence of "PRE" in this file.